### PR TITLE
[controlls] Prioritize spell targets

### PR DESCRIPTION
**Pull request by AJenbo**
Spells will now target either the appropriate highlighted actor or the
current direction.

Wish list:
- Teleport only lets you jump to the other side of a 1 unit thick wall and only if standing next to it.
- <s>Trap Disarm, Heal Other, Resurrect and Telekinesis should trigger instantly, and do not fire if no target.</s>
- Telekinesis only looks for the item next to you.

This solves multiple point on https://github.com/diasurgical/devilutionX/issues/418

### CODE REVIEW

File: SourceX/controls/plrctrls.cpp [hunk 1]
```diff
@@ -17,7 +17,14 @@ int speedspellcount = 0;
 // Native game menu, controlled by simulating a keyboard.
 bool InGameMenu()
 {
-	return stextflag > 0 || questlog || helpflag || talkflag || qtextflag || sgpCurrentMenu;
+	return stextflag > 0
+		|| questlog
+		|| helpflag
+		|| talkflag
+		|| qtextflag
+		|| gmenu_exception()
```
Comment (glebm) [inline SourceX/controls/plrctrls.cpp]: Does this replace `sgpCurrentMenu`?

File: SourceX/controls/plrctrls.cpp [hunk 2]
```diff
@@ -664,6 +671,81 @@ void PerformPrimaryAction()
...
+{
+	if (SpellHasActorTarget())
+		return;
+
+	pcursplr = -1;
+	pcursmonst = -1;
+
+	cursmx = plr[myplr]._px;
+	cursmy = plr[myplr]._py;
+	switch (plr[myplr]._pdir) {
```
Comment (glebm) [inline SourceX/controls/plrctrls.cpp]: Perhaps it would be cleaner to make a lookup table

File: SourceX/controls/plrctrls.cpp (hunk 1 above)
Comment (AJenbo) [reply SourceX/controls/plrctrls.cpp]: Yes

File: SourceX/miniwin/misc_msg.cpp [hunk 3]
```diff
@@ -406,13 +406,9 @@ WINBOOL PeekMessageA(LPMSG lpMsg, HWND hWnd, UINT wMsgFilterMin, UINT wMsgFilter
...
 		case GameActionType::CAST_SPELL:
 			if (pcurs >= CURSOR_FIRSTITEM && invflag)
-				// Drop item so that it does not get destroyed.
 				DropItemBeforeTrig();
-			if (!invflag && !talkflag)
-				// Cast the spell.
-				RightMouseDown();
-			// Close active menu if any / stop walking.
-			PressEscKey();
+			else
```
Comment (glebm) [inline SourceX/miniwin/misc_msg.cpp]: Should this be in the `else` branch? Previously the spell was cast even if the inventory was open

File: SourceX/controls/plrctrls.cpp (hunk 2 above)
Comment (AJenbo) [reply SourceX/controls/plrctrls.cpp]: Would you make a table for each axis?

File: SourceX/miniwin/misc_msg.cpp (hunk 3 above)
Comment (AJenbo) [reply SourceX/miniwin/misc_msg.cpp]: It feelt better to have only one function for the key at one time, this way you can drop items with out also casting a random firewall :)

File: SourceX/controls/plrctrls.cpp (hunk 2 above)
Comment (glebm) [reply SourceX/controls/plrctrls.cpp]: Perhaps one 2-dimensional table to fetch both values at once?

Something like this:

```c++
static const int kOffsets[8][2] = {
  {  1,  1 }, // DIR_S
  {  0,  1 }, // DIR_SW
  { -1,  1 }, // DIR_W
  { -1,  0 }, // DIR_NW
  { -1, -1 }, // DIR_N
  {  0, -1 }, // DIR_NE
  {  1, -1 }, // DIR_E
  {  1,  0 }, // DIR_SE
};
const auto &offset = kOffsets[plr[myplr]._pdir];
cursmx = plr[myplr]._px + offset[0];
cursmy = plr[myplr]._py + offset[1];
```

File: SourceX/controls/plrctrls.cpp [hunk 4]
```diff
@@ -664,15 +716,96 @@ void PerformPrimaryAction()
...
+	if (InGameMenu())
+		return;
+	if (TryIconCurs())
+		return;
+
+	int spl = plr[myplr]._pRSpell;
+	if ((pcursplr == -1 && (spl == SPL_RESURRECT || spl == SPL_HEALOTHER))
+		|| (pcursobj == -1 && spl == SPL_DISARM)) {
+		if (plr[myplr]._pClass == PC_WARRIOR) {
+			PlaySFX(PS_WARR27);
```
Comment (AJenbo) [inline SourceX/controls/plrctrls.cpp]: I cant cast that here

**Pull request merged**